scipy/interpolate/regrid_python.py

```diff
@@ -0,0 +1,1174 @@
+        a compatible shape that can be reshaped to that.
+    degrees : tuple of int
+        Degrees ``(kx, ky)`` along x and y.
+
+    Returns
+    -------
+    NdBSpline
+        The constructed 2D spline.
+
+    Notes
+    -----
+    Only repacks the coefficient grid; ``fp`` is not used internally here.
+    """
+    nx, ny = len(tck[0]), len(tck[1])
+    kx, ky = degrees
+    c = tck[2].reshape(nx - kx - 1, ny - ky - 1)
+    return NdBSpline((tck[0], tck[1]), c, degrees)
+
+
+def _stack_augmented_fitpack(A, offs_a, D, offs_d, nc, k, p):
+    if p == -1:
+        return A, offs_a, nc
+
+    nz = k + 1
+    AA = np.zeros((nc + D.shape[0], k+2), dtype=float)
+    AA[:nc, :nz] = A[:nc, :]
+    AA[nc:, :] = D / p
+    offset = np.r_[offs_a, offs_d]
+    return AA, offset, nc
+
+def will_square_overflow(x, dtype=np.float64):
+    """
+    Check if squaring values would overflow the given dtype.
+
+    Parameters
+    ----------
+    x : array_like
+        Input values to test.
+    dtype : dtype, optional
+        Floating dtype to test against, by default ``np.float64``.
+
+    Returns
+    -------
+    ndarray of bool
+        Boolean mask where True indicates ``x**2`` would overflow ``dtype``.
+
+    Notes
+    -----
+    Compares ``abs(x)`` to ``sqrt(np.finfo(dtype).max)`` elementwise.
+    """
+    max_val = np.finfo(dtype).max
+    return abs(x) > np.sqrt(max_val)
+
+def fp_residual(Z, Zhat, W=None):
+    """
+    Compute FITPACK-style residual sum of squares ``fp``.
+
+    Parameters
+    ----------
+    Z : array_like
+        Target data (observed grid).
+    Zhat : array_like
+        Predicted/fitted grid, broadcastable to ``Z``.
+    W : array_like, optional
+        Optional weights, same shape as ``Z``; if provided, compute
+        ``sum(W * (Z - Zhat)**2)``.
+
+    Returns
+    -------
+    float
+        Residual sum of squares, or ``inf`` if non-finite or overflow detected.
+
+    Notes
+    -----
+    Guards against NaNs/Infs and overflow by early returning ``inf``.
+    """
+    R = Z - Zhat
+
+    if not np.isfinite(R).all():
+        return float("inf")
+
+    if np.isnan(R).any() or will_square_overflow(R).any():
+        return float("inf")
+
+    R2 = np.square(R, dtype=np.float64)
+    if W is None:
+        return float(np.sum(R2))
+    else:
+        return float(np.sum(R2 * W))
+
+def _solve_2d_fitpack(Ax, offs_x, ncx,
+                      Dx, offs_dx, ncdx,
+                      Ay, offs_y, ncy, Q,
+                      Dy, offs_dy, ncdy, p,
+                      kx, tx, x_x, w_x,
+                      ky, ty, x_y, w_y, z):
+
+    Ax_aug, offset_aug_x, nc_augx = _stack_augmented_fitpack(
+        Ax, offs_x, Dx, offs_dx, ncx, kx, p)
+    Ay_aug, offset_aug_y, nc_augy = _stack_augmented_fitpack(
+        Ay, offs_y, Dy, offs_dy, ncy, ky, p)
+
+    if p != -1:
+        Q = np.vstack([Q, np.zeros((Dx.shape[0], Q.shape[1]), dtype=np.float64)])
```

Comment:
What is  `p==-1`?

Reply:
(Sorry for a piecemeal review). Minor readability tweak: here you have three `if p != -1` branches: one here and two in `_stack_augmented_fitpack`. Better make it branch once.